### PR TITLE
task/401 Contribution Building Block

### PR DESCRIPTION
**Pull request by ywkim312**

### CODE REVIEW

File: contributions/api/controllers/contributions.py [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+            return rs_handlers.bad_request(msg)
+    except:
+        pass
+
+    if is_new_install:
+        # new installation of the app
+        currenttime = datetime.datetime.now()
+        currenttime = currenttime.strftime("%Y/%m/%dT%H:%M:%S")
+        contribution_dataset = Contribution('')
+        contribution_dataset, restjson = datasetutils.update_capability_dataset_from_json(contribution_dataset, in_json)
```
Comment (sandeep-ps) [inline contributions/api/controllers/contributions.py:67]:
```suggestion
        contribution_dataset, restjson = datasetutils.update_contribution_dataset_from_json(contribution_dataset, in_json)
```
I think this needs to be update. Right?

File: contributions/api/controllers/contributions.py [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+from models.capabilities.capability import Capability
+from models.talents.talent import Talent
+from pymongo import MongoClient
+
+client_contribution = MongoClient(cfg.MONGO_CONTRIBUTION_URL, connect=False)
+db_contribution = client_contribution[cfg.CONTRIBUTION_DB_NAME]
+coll_contribution = db_contribution[cfg.CONTRIBUTION_COLL_NAME]
+
+def post():
+    is_new_install = True
```
Comment (sandeep-ps) [inline contributions/api/controllers/contributions.py:42]:
```suggestion
    is_new_install = True
    in_json = None
```
I think it's better to initialize in_json to None.

File: contributions/api/controllers/contributions.py [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+    }
+    msg_json = jsonutils.create_log_json("Talent", "SEARCH", msg)
+    logging.info("Talent SEARCH " + json.dumps(msg))
+
+    return talent_dataset
+
+def construct_capability(in_json):
+    is_required_field = True
+    error_required = ""
+    try:
```
Comment (sandeep-ps) [inline contributions/api/controllers/contributions.py:440]: If a required field is missing, this try block won't be executed. Connexion will return bad request error to client based on the API specification.

File: contributions/api/utils/datasetutils.py [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+        pass
+    try:
+        dataset.set_api_doc_url(injson['apiDocUrl'])
+        del outjson['apiDocUrl']
+    except:
+        pass
+    try:
+        deployment_details = DeploymentDetails()
+        try:
+            location = injson["deploymentLocation"]["location"]
```
Comment (sandeep-ps) [inline contributions/api/utils/datasetutils.py:98]:
```suggestion
            location = injson["deploymentDetails"]["location"]
```

File: contributions/api/utils/datasetutils.py [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+            deployment_details.set_location(location)
+        except:
+            pass
+        try:
+            docker_image_name = injson["deploymentLocation"]["dockerImageName"]
```
Comment (sandeep-ps) [inline contributions/api/utils/datasetutils.py:103]:
```suggestion
            docker_image_name = injson["deploymentDetails"]["dockerImageName"]
```

File: contributions/api/utils/datasetutils.py [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+            deployment_details.set_docker_image_name(docker_image_name)
+        except:
+            pass
+        try:
+            environment_variable = EnvironmentVariable()
+            key = injson["deploymentLocation"]["environmentVariable"]["key"]
```
Comment (sandeep-ps) [inline contributions/api/utils/datasetutils.py:109]:
```suggestion
            key = injson["deploymentDetails"]["environmentVariable"]["key"]
```

File: contributions/api/utils/datasetutils.py [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+            value = injson["deploymentLocation"]["environmentVariable"]["value"]
```
Comment (sandeep-ps) [inline contributions/api/utils/datasetutils.py:110]:
```suggestion
            value = injson["deploymentDetails"]["environmentVariable"]["value"]
```

File: contributions/api/utils/datasetutils.py (hunk 6 above)
Comment (sandeep-ps) [inline contributions/api/utils/datasetutils.py:109]: Environment variables is a list of dictionary containing `key`, `value` pairs. This needs to be updated.

File: contributions/api/utils/datasetutils.py [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+            environment_variable.set_key(key)
+            environment_variable.set_value(value)
+            deployment_details.set_environment_variable(environment_variable)
+            del outjson["environmentVariable"]
+        except Exception as e:
+            pass
+        try:
+            database_details = injson["deploymentLocation"]["databaseDetails"]
```
Comment (sandeep-ps) [inline contributions/api/utils/datasetutils.py:118]:
```suggestion
            database_details = injson["deploymentDetails"]["databaseDetails"]
```

File: contributions/api/utils/datasetutils.py [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+            deployment_details.set_database_details(database_details)
+        except:
+            pass
+        try:
+            auth_method = injson["deploymentLocation"]["authMethod"]
```
Comment (sandeep-ps) [inline contributions/api/utils/datasetutils.py:123]:
```suggestion
            auth_method = injson["deploymentDetails"]["authMethod"]
```

File: contributions/api/utils/datasetutils.py [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+            deployment_details.set_auth_method(auth_method)
+        except:
+            pass
+        dataset.set_deployment_details(deployment_details)
+        del outjson["deploymentLocation"]
```
Comment (sandeep-ps) [inline contributions/api/utils/datasetutils.py:128]:
```suggestion
        del outjson["deploymentDetails"]
```

File: contributions/api/controllers/contributions.py (hunk 1 above)
Comment (ywkim312) [reply contributions/api/controllers/contributions.py:67]: changed

File: contributions/api/controllers/contributions.py (hunk 2 above)
Comment (ywkim312) [reply contributions/api/controllers/contributions.py:42]: added

File: contributions/api/controllers/contributions.py (hunk 3 above)
Comment (ywkim312) [reply contributions/api/controllers/contributions.py:440]: Yes. I put this for like an insurance when the yaml was not really fixed, but I don't really remember 100 percent.

File: contributions/api/utils/datasetutils.py (hunk 4 above)
Comment (ywkim312) [reply contributions/api/utils/datasetutils.py:98]: changed

File: contributions/api/utils/datasetutils.py (hunk 5 above)
Comment (ywkim312) [reply contributions/api/utils/datasetutils.py:103]: changed

File: contributions/api/utils/datasetutils.py (hunk 6 above)
Comment (ywkim312) [reply contributions/api/utils/datasetutils.py:109]: changed

File: contributions/api/utils/datasetutils.py (hunk 7 above)
Comment (ywkim312) [reply contributions/api/utils/datasetutils.py:110]: changed

File: contributions/api/utils/datasetutils.py (hunk 8 above)
Comment (ywkim312) [reply contributions/api/utils/datasetutils.py:118]: changed

File: contributions/api/utils/datasetutils.py (hunk 9 above)
Comment (ywkim312) [reply contributions/api/utils/datasetutils.py:123]: changed

File: contributions/api/utils/datasetutils.py (hunk 10 above)
Comment (ywkim312) [reply contributions/api/utils/datasetutils.py:128]: changed

File: CHANGELOG.md [hunk 11]
```diff
@@ -77,6 +77,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [1.4.0] - 2020-06-11
 ### Added
 - Index to PII and Non-PII database collections. [#428](https://github.com/rokwire/rokwire-building-blocks-api/issues/428)
+- Add search option in contribution and capability dataset. [#458](https://github.com/rokwire/rokwire-building-blocks-api/issues/458)
```
Comment (sandeep-ps) [inline CHANGELOG.md:80]: After https://github.com/rokwire/rokwire-building-blocks-api/pull/571 gets reviewed and merged, please move the entries in this PR for Contributions catalog under the `[Unreleased]` section. Thanks.

File: contributions/contribution.yaml [hunk 12]
```diff
@@ -0,0 +1,509 @@
+openapi: 3.0.0
+info:
+  title: Rokwire Contribution Building Block API
```
Comment (sandeep-ps) [inline contributions/contribution.yaml:3]:
```suggestion
  title: Rokwire Contributions Building Block API
```

File: contributions/contribution.yaml [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+  description: Contributino Building Block API Documentation
```
Comment (sandeep-ps) [inline contributions/contribution.yaml:4]:
```suggestion
  description: Contributions Building Block API Documentation
```

File: contributions/api/controllers/configs.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+# this has to be have the file path to the folder for saving image files and other necessary files from rest service
+MONGO_CONTRIBUTION_URL = os.getenv('MONGO_CONTRIBUTION_URL', 'localhost:27017')
+FLASK_APP = os.getenv('FLASK_APP', 'contribution_rest_service')
+FLASK_ENV = os.getenv('FLASK_ENV', 'production')
+API_LOC = os.getenv('API_LOC', '../')
+CONTRIBUTION_URL_PREFIX = os.getenv('CONTRIBUTION_URL_PREFIX', '')
+DEBUG = bool(os.getenv('DEBUG', 'False') == 'True')
+
+CONTRIBUTION_DB_NAME = 'contributiondb'
```
Comment (sandeep-ps) [inline contributions/api/controllers/configs.py:30]:
```suggestion
CONTRIBUTION_DB_NAME = 'contributions_db'
```

Also, please make CONTRIBUTION_DB_NAME, CONTRIBUTION_COLL_NAME, CAPABILITY_COLL_NAME configurable via environment variables.

File: CHANGELOG.md (hunk 11 above)
Comment (ywkim312) [reply CHANGELOG.md]: updated

File: contributions/README.md [hunk 15]
```diff
@@ -0,0 +1,1015 @@
+# Contribution Building Block
+
+**contribution rest service** is a Python project to provide rest service for rokwire building block contribution
```
Comment (sandeep-ps) [inline contributions/README.md:3]:
```suggestion
Contributions Building Block provides a set of RESTFul web services to manage Contributions in the Rokwire platform. This includes adding a new contribution (a combination of Talents and Capabilities), retrieving relevant contributions based on search query, updating and deleting existing contributions, etc.
```

File: contributions/README.md [hunk 16]
```diff
@@ -0,0 +1,1015 @@
+# Contribution Building Block
```
Comment (sandeep-ps) [inline contributions/README.md:1]:
```suggestion
# Contributions Building Block
```

Comment (ywkim312) [reply contributions/README.md]: changed

File: contributions/README.md (hunk 15 above)
Comment (ywkim312) [reply contributions/README.md]: changed

File: contributions/api/controllers/configs.py (hunk 14 above)
Comment (ywkim312) [reply contributions/api/controllers/configs.py]: changed

File: contributions/contribution.yaml (hunk 13 above)
Comment (ywkim312) [reply contributions/contribution.yaml]: changed

File: contributions/contribution.yaml (hunk 12 above)
Comment (ywkim312) [reply contributions/contribution.yaml]: changed

File: contributions/api/utils/jsonutils.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+from flask import make_response
+
+def remove_objectid_from_dataset(dataset):
+    if "_id" in dataset:
+        del dataset["_id"]
+
+    return dataset
+
+def create_log_json(ep_name, ep_method, in_json):
+    in_json['ep_building_block'] = "contribution_building_block"
```
Comment (sandeep-ps) [inline contributions/api/utils/jsonutils.py:24]:
```suggestion
    in_json['ep_building_block'] = "contributions_building_block"
```

File: contributions/api/utils/query_params.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from bson import ObjectId
+
+
+def format_query(args, query):
```
Comment (sandeep-ps) [inline contributions/api/utils/query_params.py:18]: This method can be removed, right?

File: CHANGELOG.md (hunk 11 above)
Comment (sandeep-ps) [reply CHANGELOG.md]: Thanks.

File: contributions/api/utils/query_params.py (hunk 18 above)
Comment (ywkim312) [reply contributions/api/utils/query_params.py]: removed

File: contributions/api/utils/jsonutils.py (hunk 17 above)
Comment (ywkim312) [reply contributions/api/utils/jsonutils.py]: removed

**Review by sandeep-ps — APPROVED** [review]
Thanks! Approving and merging this.

**Pull request merged**